### PR TITLE
Feature/as 4245 has date decision due

### PR DESCRIPTION
**Pull request by swdpcomputing**
## Ticket Number

https://pins-ds.atlassian.net/browse/AS-4245

## Description of change

add has date decision due page

## Checklist

- [ ] Requires infrastructure changes
- [ ] If adding or remove environment variables (e.g. in `docker-compose.yaml`) then I have updated the appropriate Helm chart
- [ ] I have updated the documentation accordingly
- [x] My commit history in this PR is linear
- [x] New features have tests
- [ ] Breaking change (team conversation required)

## Important

Please do not merge from `master` (please only [rebase](https://github.com/foundry4/appeal-planning-decision/wiki/An-intro-to-Git-Rebase)). This keeps the history linear and easier to debug.

### CODE REVIEW

File: packages/forms-web-app/src/validators/householder-planning/eligibility/date-decision-due-householder.js [hunk 1]
```diff
@@ -0,0 +1,24 @@
+const { body } = require('express-validator');
+const { isAfter } = require('date-fns');
+const dateInputValidation = require('../../custom/date-input');
+
+const rules = () => [
+  ...(dateInputValidation('date-decision-due-householder', 'the Decision Date') || []),
```
Comment (nssdesikan) [inline packages/forms-web-app/src/validators/householder-planning/eligibility/date-decision-due-householder.js:6]: Is this the correct error message?

File: packages/e2e-tests/cypress/support/eligibility/granted-or-refused-application/select-planning-application-decision.js [hunk 2]
```diff
@@ -1,14 +1,16 @@
-import {getGrantedRadio, getRefusedRadio, getNoDecisionReceivedRadio, getPlanningApplicationDecisionError} from '../page-objects/granted-or-refused-application-po'
+import {getGrantedRadio, getRefusedRadio, getNoDecisionReceivedRadio
+} from '../page-objects/granted-or-refused-application-po'
 
 export const selectPlanningApplicationDecision = (decision)=>{
+  console.log('aaaaaaa '+decision)
```
Comment (demzzydev) [inline packages/e2e-tests/cypress/support/eligibility/granted-or-refused-application/select-planning-application-decision.js:5]: remove console log

File: packages/e2e-tests/cypress/support/eligibility/granted-or-refused-application/select-planning-application-decision.js [hunk 3]
```diff
@@ -1,14 +1,16 @@
...
     switch(decision){
         case 'Granted':
```
Comment (demzzydev) [inline packages/e2e-tests/cypress/support/eligibility/granted-or-refused-application/select-planning-application-decision.js:7]: Might be a better idea to use constants

File: packages/forms-web-app/__tests__/unit/routes/householder-planning/eligibility/index.test.js [hunk 4]
```diff
@@ -14,9 +14,9 @@ describe('routes/householder-planning/eligibility/index', () => {
 
   it('should define the expected routes', () => {
     expect(use).toHaveBeenCalledWith(listedBuildingHouseholderRouter);
-    expect(use).toHaveBeenCalledWith(enforcementNoticeHouseholderRouter);
-    expect(use).toHaveBeenCalledWith(grantedOrRefusedHouseholderRouter);
+    expect(use).toHaveBeenCalledWith(dateDecisionDueHouseholderRouter);
+    expect(use).toHaveBeenCalledWith(grantedOrRefusedRouter);
 
-    expect(use.mock.calls.length).toBe(3);
+    expect(use.mock.calls.length).toBe(4);
```
Comment (demzzydev) [inline packages/forms-web-app/__tests__/unit/routes/householder-planning/eligibility/index.test.js:20]: Where's the fourth router?

**Review by nssdesikan — CHANGES_REQUESTED** [review]

File: packages/e2e-tests/cypress/support/eligibility/granted-or-refused-application/select-planning-application-decision.js (hunk 2 above)
Comment (swdpcomputing) [reply packages/e2e-tests/cypress/support/eligibility/granted-or-refused-application/select-planning-application-decision.js:5]: Removed.

File: packages/e2e-tests/cypress/support/eligibility/granted-or-refused-application/select-planning-application-decision.js (hunk 3 above)
Comment (swdpcomputing) [reply packages/e2e-tests/cypress/support/eligibility/granted-or-refused-application/select-planning-application-decision.js:7]: This is Shilpa's domain so going to leave as is. I think it's fine for a short switch.

File: packages/forms-web-app/__tests__/unit/routes/householder-planning/eligibility/index.test.js (hunk 4 above)
Comment (swdpcomputing) [reply packages/forms-web-app/__tests__/unit/routes/householder-planning/eligibility/index.test.js:20]: listed-building, granted-or-refused, enforcement-notice, date-decision-due

File: packages/forms-web-app/src/views/householder-planning/eligibility/date-decision-due-householder.njk [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+                id: "date-decision-due-householder-year",
+                value: decisionDate.year
+              }
+            ]
+          }) }}
+        </div>
+
+        {{ govukButton({
+          text: "Continue",
+          type: "Submit",
```
[the remark this answers is not in the document]
Comment (swdpcomputing) [reply packages/forms-web-app/src/views/householder-planning/eligibility/date-decision-due-householder.njk:83]: Edited

File: packages/forms-web-app/src/views/householder-planning/eligibility/date-decision-due-householder.njk [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+      <form action="" method="POST" novalidate>
+        <div class="govuk-form-group {{ 'govuk-form-group--error' if errors }}">
+          <span class="govuk-caption-l">Before you start</span>
+
+          {{ govukDateInput({
+            id: "date-decision-due-householder",
+            namePrefix: "date-decision-due-householder",
+            fieldset: {
+              legend: {
+                text: "What date was the decision due?​",
```
[the remark this answers is not in the document]
Comment (swdpcomputing) [reply packages/forms-web-app/src/views/householder-planning/eligibility/date-decision-due-householder.njk:48]: Ticket takes priority.

File: packages/forms-web-app/src/controllers/householder-planning/eligibility/date-decision-due-householder.js [hunk 7]
```diff
@@ -0,0 +1,57 @@
+const { add, isBefore } = require('date-fns');
+const logger = require('../../../lib/logger');
+const { createOrUpdateAppeal } = require('../../../lib/appeals-api-wrapper');
+const { VIEW } = require('../../../lib/householder-planning/views');
+
+exports.getDateDecisionDueHouseholder = async (req, res) => {
+  res.render(VIEW.HOUSEHOLDER_PLANNING.ELIGIBILITY.DATE_DECISION_DUE_HOUSEHOLDER, {
+    backLink: `/before-you-start/granted-or-refused-householder`,
```
[the remark this answers is not in the document]
Comment (swdpcomputing) [reply packages/forms-web-app/src/controllers/householder-planning/eligibility/date-decision-due-householder.js:8]: Edited.

File: packages/forms-web-app/src/controllers/householder-planning/eligibility/date-decision-due-householder.js [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+  }
+
+  const todaysDate = Date.now();
+  const enteredDate = new Date(
+    body['date-decision-due-householder-year'],
+    (parseInt(body['date-decision-due-householder-month'], 10) - 1).toString(),
+    body['date-decision-due-householder-day']
+  );
+
+  if (isBefore(enteredDate, add(new Date(todaysDate), { months: -6, days: -1 }))) {
```
[the remark this answers is not in the document]
Comment (swdpcomputing) [reply packages/forms-web-app/src/controllers/householder-planning/eligibility/date-decision-due-householder.js:37]: We can review this later. For speed at the moment, leaving as is.

File: packages/forms-web-app/src/controllers/householder-planning/eligibility/date-decision-due-householder.js [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+    return res.redirect('/before-you-start/you-cannot-appeal');
```
[the remark this answers is not in the document]
Comment (swdpcomputing) [reply packages/forms-web-app/src/controllers/householder-planning/eligibility/date-decision-due-householder.js:38]: Done.

File: packages/forms-web-app/src/routes/householder-planning/index.js [hunk 10]
```diff
@@ -1,3 +1,4 @@
+/* eslint-disable import/no-unresolved */
```
[the remark this answers is not in the document]
Comment (swdpcomputing) [reply packages/forms-web-app/src/routes/householder-planning/index.js:1]: Removed

File: packages/forms-web-app/__tests__/unit/routes/householder-planning/eligibility/index.test.js [hunk 11]
```diff
@@ -1,8 +1,8 @@
 const { use } = require('../../router-mock');
 
 const listedBuildingHouseholderRouter = require('../../../../../src/routes/householder-planning/eligibility/listed-building-householder');
-const enforcementNoticeHouseholderRouter = require('../../../../../src/routes/householder-planning/eligibility/enforcement-notice-householder');
-const grantedOrRefusedHouseholderRouter = require('../../../../../src/routes/householder-planning/eligibility/granted-or-refused-householder');
+const dateDecisionDueHouseholderRouter = require('../../../../../src/routes/householder-planning/eligibility/date-decision-due-householder');
```
[the remark this answers is not in the document]
Comment (swdpcomputing) [reply packages/forms-web-app/__tests__/unit/routes/householder-planning/eligibility/index.test.js:4]: Added.

File: packages/forms-web-app/src/validators/householder-planning/eligibility/date-decision-due-householder.js (hunk 1 above)
Comment (swdpcomputing) [reply packages/forms-web-app/src/validators/householder-planning/eligibility/date-decision-due-householder.js:6]: Yes, Shahid has not edited the ticket yet to reflect GDS

File: packages/forms-web-app/__tests__/unit/controllers/householder-planning/eligibility/listed-building-householder.test.js [hunk 12]
```diff
@@ -60,7 +59,7 @@ describe('controllers/householder-planning/listed-building-householder', () => {
         ...appeal,
       });
 
-      expect(res.redirect).toBeCalledWith('/before-you-start/granted-or-refused-householder');
+      expect(res.redirect).toBeCalledWith('/before-you-start/enforcement-notice-householder'); // Future Planning Application Decision Page
```
[the remark this answers is not in the document]
Comment (swdpcomputing) [reply packages/forms-web-app/__tests__/unit/controllers/householder-planning/eligibility/listed-building-householder.test.js:62]: Removed.

File: packages/e2e-tests/cypress/support/eligibility/granted-or-refused-application/select-planning-application-decision.js [hunk 13]
```diff
@@ -1,14 +1,16 @@
...
             getGrantedRadio().check();
             break;
         case 'Refused':
             getRefusedRadio().check();
             break;
-        case 'I have not received a decision':
+        case 'I have Not Received a Decision':
```
[the remark this answers is not in the document]
Comment (swdpcomputing) [reply packages/e2e-tests/cypress/support/eligibility/granted-or-refused-application/select-planning-application-decision.js:13]: For Shilpa, see if it passes tests

File: packages/forms-web-app/src/controllers/householder-planning/eligibility/date-decision-due-householder.js [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+    });
+    return res.redirect(`/before-you-start/enforcement-notice-householder`);
+  } catch (e) {
+    logger.error(e);
+
+    return res.render(VIEW.HOUSEHOLDER_PLANNING.ELIGIBILITY.DATE_DECISION_DUE_HOUSEHOLDER, {
+      appeal,
+      errors,
+      errorSummary: [{ text: e.toString(), href: 'date-decision-due-householder' }],
+      backLink: `/before-you-start/granted-or-refused-householder`,
```
[the remark this answers is not in the document]
Comment (swdpcomputing) [reply packages/forms-web-app/src/controllers/householder-planning/eligibility/date-decision-due-householder.js]: Done.

File: packages/forms-web-app/src/controllers/householder-planning/eligibility/date-decision-due-householder.js [hunk 15]
```diff
@@ -0,0 +1,57 @@
...
+  if (isBefore(enteredDate, add(new Date(todaysDate), { months: -6, days: -1 }))) {
+    return res.redirect('/before-you-start/you-cannot-appeal');
+  }
+
+  try {
+    req.session.appeal = await createOrUpdateAppeal({
+      ...appeal,
+      decisionDate: enteredDate.toISOString(),
+    });
+    return res.redirect(`/before-you-start/enforcement-notice-householder`);
```
[the remark this answers is not in the document]
Comment (swdpcomputing) [reply packages/forms-web-app/src/controllers/householder-planning/eligibility/date-decision-due-householder.js]: Done.

File: packages/forms-web-app/__tests__/unit/controllers/householder-planning/eligibility/date-decision-due-householder.test.js [hunk 16]
```diff
@@ -0,0 +1,141 @@
...
+      const mockRequest = {
+        ...req,
+        body: {
+          'date-decision-due-householder-year': '2021',
+          'date-decision-due-householder-month': '01',
+          'date-decision-due-householder-day': '01',
+        },
+      };
+
+      global.Date.now = jest.fn(() => new Date('2021-02-01T00:00:00.000Z').getTime());
```
[the remark this answers is not in the document]
Comment (swdpcomputing) [reply packages/forms-web-app/__tests__/unit/controllers/householder-planning/eligibility/date-decision-due-householder.test.js:58]: So we can explicitly test exact time differences.

File: packages/forms-web-app/__tests__/unit/controllers/householder-planning/eligibility/date-decision-due-householder.test.js [hunk 17]
```diff
@@ -0,0 +1,141 @@
...
+        }
+      );
+    });
+  });
+
+  describe('postDateDecisionDueHouseholder', () => {
+    it('should save the appeal and redirect to enforcement-notice if date is within six months', async () => {
+      const mockRequest = {
+        ...req,
+        body: {
```
[the remark this answers is not in the document]
Comment (swdpcomputing) [reply packages/forms-web-app/__tests__/unit/controllers/householder-planning/eligibility/date-decision-due-householder.test.js:51]: Yes, they're hardcoded.

File: packages/forms-web-app/__tests__/unit/controllers/householder-planning/eligibility/date-decision-due-householder.test.js [hunk 18]
```diff
@@ -0,0 +1,141 @@
...
+          'date-decision-due-householder-year': '2021',
+          'date-decision-due-householder-month': '01',
+          'date-decision-due-householder-day': '01',
+        },
+      };
+
+      global.Date.now = jest.fn(() => new Date('2021-10-01T00:00:00.000Z').getTime());
```
[the remark this answers is not in the document]
Comment (swdpcomputing) [reply packages/forms-web-app/__tests__/unit/controllers/householder-planning/eligibility/date-decision-due-householder.test.js:80]: So we can explicitly test exact time differences.

File: packages/forms-web-app/src/views/householder-planning/eligibility/date-decision-due-householder.njk (hunk 6 above)
[the remark this answers is not in the document]
Comment (nssdesikan) [reply packages/forms-web-app/src/views/householder-planning/eligibility/date-decision-due-householder.njk:48]: Text different to the story.

```suggestion
                text: "What date was your decision due?​",
```

File: packages/forms-web-app/src/controllers/householder-planning/eligibility/date-decision-due-householder.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (nssdesikan) [reply packages/forms-web-app/src/controllers/householder-planning/eligibility/date-decision-due-householder.js:41]: Can you create a placeholder card for this future work please?

File: packages/forms-web-app/src/validators/householder-planning/eligibility/date-decision-due-householder.js (hunk 1 above)
Comment (nssdesikan) [reply packages/forms-web-app/src/validators/householder-planning/eligibility/date-decision-due-householder.js:6]: The test ix clearly wrong. Do you want me to chase Shahid or you want to chase him?

File: packages/e2e-tests/cypress/support/eligibility/granted-or-refused-application/select-planning-application-decision.js (hunk 13 above)
[the remark this answers is not in the document]
Comment (nssdesikan) [reply packages/e2e-tests/cypress/support/eligibility/granted-or-refused-application/select-planning-application-decision.js:12]: @shilpabasu22 Can you answer to this please?

[the remark this answers is not in the document]
Comment (shilpabasu22) [reply packages/e2e-tests/cypress/support/eligibility/granted-or-refused-application/select-planning-application-decision.js:12]: This is what was returned by the application and the original one was failing the tests, I think, we can have another case here

[the remark this answers is not in the document]
Comment (shilpabasu22) [reply packages/e2e-tests/cypress/support/eligibility/granted-or-refused-application/select-planning-application-decision.js:12]: I have fixed this.

**Review by nssdesikan — DISMISSED** [review]

**Review by nssdesikan — APPROVED** [review]

**Pull request merged**